### PR TITLE
chore: update to mathlib 20/05/2024

### PR DESCRIPTION
**Pull request by tobiasgrosser**
by running `lake -Kdoc=on update`

### CODE REVIEW

File: SSA/Projects/DCE/DCE.lean [hunk 1]
```diff
@@ -27,22 +25,6 @@ theorem List.eraseIdx_eq_len_concat : List.eraseIdx (xs ++ [x]) xs.length = xs :
...
-    case cons x xs' IH => simp at hn
-  case succ n' IH' =>
-    induction xs
-    case nil => simp[eraseIdx]
-    case cons x xs' IH =>
-      simp[eraseIdx_succ]
-      apply IH'
-      simp at hn
-      linarith
-
```
Comment (bollu) [inline SSA/Projects/DCE/DCE.lean:45]: Why is this deleted?

Comment (tobiasgrosser) [reply SSA/Projects/DCE/DCE.lean:45]: This is now in upstream.

File: SSA/Projects/FullyHomomorphicEncryption/Basic.lean [hunk 2]
```diff
@@ -466,7 +466,7 @@ theorem R.fromTensor_eq_fromTensor'_fromPoly_aux (coeffs : List Int) (rp : R q n
 /-- fromTensor = R.fromPoly ∘ fromTensor'.
 This permits reasoning about fromTensor directly on the polynomial ring.
 -/
-theorem R.fromTensor_eq_fromTensor'_fromPoly {q n} : R.fromTensor (q := q) (n := n) coeffs =
+theorem R.fromTensor_eq_fromTensor'_fromPoly {q n} {coeffs : List Int} : R.fromTensor (q := q) (n := n) coeffs =
```
[the remark this answers is not in the document]
Comment (tobiasgrosser) [reply SSA/Projects/FullyHomomorphicEncryption/Basic.lean:469]: It derived the wrong type.

File: SSA/Projects/FullyHomomorphicEncryption/Basic.lean [hunk 3]
```diff
@@ -545,7 +545,7 @@ theorem R.fromTensorFinsupp_concat_monomial {q : ℕ} (c : ℤ) (cs : List ℤ)
     rw[List.length_map]
 
 /-- show that `fromTensor` is the same as `fromPoly ∘ fromTensorFinsupp`. -/
-theorem R.fromTensor_eq_fromTensorFinsupp_fromPoly {q n} : R.fromTensor (q := q) (n := n) coeffs =
+theorem R.fromTensor_eq_fromTensorFinsupp_fromPoly {q n} {coeffs : List ℤ} : R.fromTensor (q := q) (n := n) coeffs =
```
[the remark this answers is not in the document]
Comment (tobiasgrosser) [reply SSA/Projects/FullyHomomorphicEncryption/Basic.lean:548]: This was previously pretty inconsistent, but I cleaned it up in: https://github.com/opencompl/ssa/pull/312
